sycl/include/CL/sycl/buffer.hpp

```diff
@@ -278,21 +255,68 @@ class buffer {
             access::target target, access::placeholder isPlaceholder>
   friend class accessor;
   range<dimensions> Range;
-  // If this buffer is subbuffer - this range represents range of the parent
+  // Offset field specifies the origin of the sub buffer inside the parent
   // buffer
-  range<dimensions> MemRange;
+  size_t OffsetInBytes = 0;
   bool IsSubBuffer = false;
-  // If this buffer is sub-buffer - offset field specifies the origin of the
-  // sub-buffer inside the parent buffer
-  id<dimensions> Offset;
 
   // Reinterpret contructor
   buffer(shared_ptr_class<detail::buffer_impl<AllocatorT>> Impl,
-         range<dimensions> reinterpretRange,
-         range<dimensions> reinterpretMemRange,
-         id<dimensions> reinterpretOffset, bool isSubBuffer)
-      : impl(Impl), Range(reinterpretRange), MemRange(reinterpretMemRange),
-        IsSubBuffer(isSubBuffer), Offset(reinterpretOffset){};
+         range<dimensions> reinterpretRange, size_t reinterpretOffset,
+         bool isSubBuffer)
+      : impl(Impl), Range(reinterpretRange), OffsetInBytes(reinterpretOffset),
+        IsSubBuffer(isSubBuffer){};
+
+  template <typename Type, int N>
+  size_t getOffsetInBytes(const id<N> &offset, const range<N> &range) {
+    return detail::getLinearIndex(offset, range) * sizeof(Type);
+  }
+
+  bool isOutOfBounds(const id<dimensions> &offset,
+                     const range<dimensions> &newRange,
+                     const range<dimensions> &parentRange) {
+    bool outOfBounds = false;
+    for (int i = 0; i < dimensions; ++i)
+      outOfBounds |= newRange[i] + offset[i] > parentRange[i];
+
+    return outOfBounds;
+  }
+
+  bool isContiguousRegion(const id<1> &offset, const range<1> &newRange,
+                          const range<1> &parentRange) {
+    // 1D sub buffer always has contiguous region
+    return true;
+  }
+
+  bool isContiguousRegion(const id<2> &offset, const range<2> &newRange,
+                          const range<2> &parentRange) {
+    // For 2D sub buffer there are 2 cases:
+    // 1) Offset {Any, Any}  | a peace of any line of a buffer
```

Comment:
Did you mean "piece"?

Reply:
Yeah, typo, will fix. Thanks for noticing.